syntax = "proto3";

package quintet.rpc;

message LogEntry {
  string opName = 1;
  string args = 2;
  uint64 term = 3;
}

// AppendEntries RPC
message AppendEntriesMessage {
  uint64 term = 1;
  string leaderId = 2;
  uint64 prevLogIndex = 3;
  uint64 prevLogTerm = 4;
  repeated LogEntry entries = 5;
  uint64 leaderCommit = 6;

  uint64 debugId = 9;
}

message AppendEntriesReply {
  uint64 term = 1;
  bool success = 2;

  uint64 debugId = 9;
}

// RequestVote RPC
message RequestVoteMessage {
  uint64 term = 1;
  string candidateId = 2;
  uint64 lastLogIndex = 3;
  uint64 lastLogTerm = 4;

  uint64 debugId = 9;
}

message RequestVoteReply {
  uint64 term = 1;
  bool voteGranted = 2;

  uint64 debugId = 9;
}

service Internal {
  rpc AppendEntries (AppendEntriesMessage) returns (AppendEntriesReply) {}

  rpc RequestVote (RequestVoteMessage) returns (RequestVoteReply) {}
}
